src/inference.rs

```diff
@@ -13,82 +10,176 @@ use crate::dataset::FSRSBatcher;
+        let time_history =
+            Tensor::<B, 1>::from_data(Data::new(time_history, Shape { dims: [size] }).convert())
+                .unsqueeze()
+                .transpose();
+        let rating_history =
+            Tensor::<B, 1>::from_data(Data::new(rating_history, Shape { dims: [size] }).convert())
+                .unsqueeze()
+                .transpose();
+        self.model().forward(time_history, rating_history).into()
+    }
+
+    pub fn next_states(
+        &self,
+        state: Option<MemoryState>,
+        desired_retention: f32,
+        days_elapsed: u32,
+    ) -> NextStates {
+        let delta_t =
+            Tensor::<B, 1>::from_data(Data::new(vec![days_elapsed as f32], Shape { dims: [1] }))
+                .unsqueeze()
+                .transpose();
+        let state = state.map(MemoryStateTensors::from);
+        let model = self.model();
+        let mut next_states = (1..=4).map(|rating| {
+            MemoryState::from(
+                model.step(
+                    delta_t.clone(),
+                    Tensor::<B, 1>::from_data(Data::new(vec![rating as f32], Shape { dims: [1] }))
+                        .unsqueeze()
+                        .transpose(),
+                    state.clone(),
+                ),
+            )
+        });
+
+        let mut get_next_state = || {
+            let memory = next_states.next().unwrap();
+            let interval = next_interval(memory.stability, desired_retention);
+            ItemState { memory, interval }
+        };
+
+        NextStates {
+            again: get_next_state(),
+            hard: get_next_state(),
+            good: get_next_state(),
+            easy: get_next_state(),
+        }
+    }
+
+    pub fn evaluate<F>(&self, items: Vec<FSRSItem>, mut progress: F) -> Result<ModelEvaluation>
+    where
+        F: FnMut(ItemProgress) -> bool,
+    {
+        let batcher = FSRSBatcher::new(self.device());
+        let mut all_predictions = vec![];
+        let mut all_true_val = vec![];
+        let mut all_retention = vec![];
+        let mut all_labels = vec![];
+        let mut progress_info = ItemProgress {
+            current: 0,
+            total: items.len(),
+        };
+        let model = self.model();
+        for chunk in items.chunks(512) {
+            let batch = batcher.batch(chunk.to_vec());
+            let (_state, retention) = infer::<B>(model, batch.clone());
+            let pred = retention.clone().squeeze::<1>(1).to_data().value;
+            all_predictions.extend(pred);
+            let true_val = batch.labels.clone().float().to_data().value;
+            all_true_val.extend(true_val);
+            all_retention.push(retention);
+            all_labels.push(batch.labels);
+            progress_info.current += chunk.len();
+            if !progress(progress_info) {
+                return Err(FSRSError::Interrupted);
+            }
+        }
+        let rmse = calibration_rmse(all_predictions, all_true_val);
```

Comment:
To avoid confusion with the common rmse, I recommend renaming `rmse` to `rmse_bins`. And keep its consistency in Anki (RMSE -> RMSE(bins)).